README.md
Torpedo
=======

Description
-----------

Fire when ready. Fast Ruby integration tests for OpenStack.

Installation
------------

1. Install the Gem.
2. Create the torpedo.conf file in your HOME directory.
3. Source the rc file for your OpenStack account (keystone credentials, etc).

```bash
	gem install torpedo

    cat > ~/.torpedo.conf <<"EOF_CAT"
	# YAML config file for torpedo

	# timeouts
	server_build_timeout: 420
	ping_timeout: 60
	ssh_timeout: 60

	# SERVER test settings
	test_create_image: false
	test_rebuild_server: false
	test_resize_server: false
	test_revert_resize_server: false
	test_admin_password: false
	test_soft_reboot_server: false
	test_hard_reboot_server: false

	# IMAGES (Set one of the following)
	image_name: Ubuntu Natty (11.04)
	#image_ref:

	# FLAVORS (Set one of the following)
	#flavor_name: 
	flavor_ref: 4

	# SSH/PING test options
	#test_ssh: true
	#test_ping: true

	# SSH KEYS (used to verify installations which support personalities)
	#ssh_private_key: <your home dir>/.ssh/id_rsa
	#ssh_public_key: <your home dir>/.ssh/id_rsa.pub

	# KEYPAIRS (used to verify images that support keypairs)
	#keypair: test.pem
	#keyname: test
	EOF_CAT

	source $PATH_TO_YOUR/openstackrc
```

Examples
--------

Available torpedo tasks:

	Tasks:
	  torpedo all          # Run all tests.
	  torpedo fire         # Fire away! (alias for all)
	  torpedo flavors      # Run flavors tests for the OSAPI.
	  torpedo help [TASK]  # Describe available tasks or one specific task
	  torpedo images       # Run images tests for the OSAPI.
	  torpedo limits       # Run limits tests for the OSAPI.
	  torpedo servers      # Run servers tests for the OSAPI.

Run all tests:

	torpedo fire

Run all tests with debug HTTP request response output:

	DEBUG=true torpedo fire

Payload
--------

* list flavors
* get flavor
* list images
* get image
* list limits
* create server (ping and ssh test w/ admin password and personality)
* delete server metadata items
* update one server metadata item
* update multiple server metadata items
* set server metadata items
* clear server metadata
* create image
* rebuild server (ping and ssh test)
* resize server (ping and ssh test)
* resize confirm (ping and ssh test)
* resize revert (ping and ssh test)
* soft reboot (ping and ssh test)
* hard reboot (ping and ssh test)
* delete image metadata
* update one image metadata item
* update multiple image metadata items
* set image metadata items
* clear image metadata

License
-------
Copyright (c) 2011-2013 Dan Prince. See LICENSE.txt for further details.


CHANGELOG
* Fri Oct 10 2013 Dan Prince <dprince@redhat.com> - 2.1.0
  -Switch from openstack-compute to fog.
  -Add 'network_namespace' option which can be used when testing
   Neutron namespaces.
  -Add support for volumes.
  -Add support for keypairs.
  -Add support for metering.
  -Add support for orchestration.

* Thu Feb 14 2013 Dan Prince <dprince@redhat.com> - 1.0.19
  -Added NETWORK_LABEL option. (defaults to 'public')

* Mon Nov 12 2012 Dan Prince <dprince@redhat.com> - 1.0.18
  -Fixes to host_id validation on resize/resize_revert tests.

* Mon Apr 9 2012 Dan Prince <dprince@redhat.com> - 1.0.17
  -Adding SLEEP_AFTER_IMAGE_CREATE env option.
  -Added 'test_limits' option.
  -Added 'test_ip_type' Set to '4' (for IPv4) or '6' (for IPv6).

* Mon Apr 2 2012 Dan Prince <dprince@redhat.com> - 1.0.16
  -Add NOVA_SERVICE_NAME env setting (defaults is nil)
  -Add NOVA_SERVICE_TYPE env setting (default is 'compute')

* Fri Mar 16 2012 Dan Prince <dprince@redhat.com> - 1.0.15
  -Set auth_method = 'rax-kskey' if NOVA_RAX_AUTH is set.

* Fri Feb 24 2012 Dan Prince <dprince@redhat.com> - 1.0.14
  -Update OS_ variables to match Openstack standard.
   OS_AUTH_KEY is now OS_PASSWORD
   OS_AUTH_USER is now OS_USERNAME
   OS_AUTH_TENANT is now OS_TENANT_NAME

* Fri Jan 6 2012 Dan Prince <dan.prince@rackspace.com> - 1.0.13
  -Update image check to allow a status other than 'ACTIVE'. Fixes
   issues when running concurrent Torpedo jobs.

* Fri Dec 30 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.12
  -Retry ssh on Errno::ECONNRESET.
  -Added cleanup task.

* Wed Dec 14 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.11
  -Set paranoid to false for ssh connection tests. (net-ssh)
  -Add options to disable server and image cleanup.
  -Remove sort on image id (we now use UUID's)
  -Add setup test to initialize image and flavor refs.

* Wed Dec 7 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.10
  -Add admin password and change password tests.
  -Add revert resize test.
  -Add hard and soft reboot tests.

* Wed Nov 16 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.9
  -Add options to disable ssh and ping tests.

* Fri Nov 11 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.8
  -Use the latest version of the openstack-compute unless a specific
   version is specified in the config file.

* Wed Nov 2 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.7
  -Add support for NOVA_REGION_NAME and NOVA_PROJECT_ID.
  -Remove dependency on test-unit-ext.
  -Add support for OS_AUTH_? style variables.

* Tue Nov 1 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.6
  -Minor metadata test updates.
  -Don't require test-unit (breaks Ruby 1.8.7)

* Tue Nov 1 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.5
  -Require test-unit if using Ruby 1.9.2.
  -Disable test-unit-ext if using Ruby 1.9.2 (currently supports Ruby 1.8.x)

* Sat Oct 29 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.4
  -Run image metadata tests last because they make the snapshot un-bootable.
  -Fixes to the resize tests.

* Sat Oct 29 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.3
  - Fix issues boolean config file issue with test_create_image,
   test_rebuild_server, and test_resize_server config options.
  - Fix a syntax issue in the get_flavor_ref_resize method.

* Sat Oct 29 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.2
  - Add missing dependency to gemspec for test-unit-ext.

* Sat Oct 29 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.1
  - Add config options to support custom flavor settings.

* Fri Oct 28 2011 Dan Prince <dan.prince@rackspace.com> - 1.0.0
  - Initial release.


lib/torpedo.rb
require 'rubygems'
require 'torpedo/config'

if RUBY_VERSION =~ /^1.9.*/ then
  gem 'test-unit'
end
require 'test/unit'
require 'test/unit/ui/console/testrunner'

configs = Torpedo::Config.load_configs

SSH_TIMEOUT=(configs['ssh_timeout'] || 30).to_i
TEST_SSH=configs.fetch('test_ssh', true)
PING_TIMEOUT=(configs['ping_timeout'] || 60).to_i
TEST_PING=configs.fetch('test_ping', true)
SERVER_BUILD_TIMEOUT=(configs['server_build_timeout'] || 60).to_i
NETWORK_LABEL=(configs['network_label'] || 'public')
NETWORK_NAMESPACE=configs['network_namespace']
SLEEP_AFTER_IMAGE_CREATE=(configs['sleep_after_image_create'] || 0).to_i
SSH_PRIVATE_KEY=configs['ssh_private_key'] || ENV['HOME'] + "/.ssh/id_rsa"
SSH_PUBLIC_KEY=configs['ssh_public_key'] || ENV['HOME'] + "/.ssh/id_rsa.pub"
TEST_CREATE_IMAGE=configs['test_create_image'] || false
TEST_REBUILD_SERVER=configs['test_rebuild_server'] || false
TEST_SOFT_REBOOT_SERVER=configs['test_soft_reboot_server'] || false
TEST_HARD_REBOOT_SERVER=configs['test_hard_reboot_server'] || false
TEST_RESIZE_SERVER=configs['test_resize_server'] || false
TEST_REVERT_RESIZE_SERVER=configs['test_revert_resize_server'] || false
TEST_ADMIN_PASSWORD=configs['test_admin_password'] || false
TEST_HOSTID_ON_RESIZE=configs['test_hostid_on_resize'] || false
TEST_IP_TYPE=configs['test_ip_type'] || 4
TEST_LIMITS=configs.fetch('test_limits', true)
CLEAN_UP_SERVERS=configs.fetch('clean_up_servers', true)
CLEAN_UP_IMAGES=configs.fetch('clean_up_images', true)
CLEAN_UP_KEYPAIRS=configs.fetch('clean_up_keypairs', true)
KEYPAIR_ENABLED = configs.fetch('keypairs', false)

IMAGE_REF=configs['image_ref']
IMAGE_NAME=configs['image_name']

FLAVOR_REF=configs['flavor_ref']
FLAVOR_NAME=configs['flavor_name']

FLAVOR_REF_RESIZE=configs['flavor_ref_resize']
FLAVOR_NAME_RESIZE=configs['flavor_name_resize']

#volume opts
volume_opts=configs['volumes'] || {}
VOLUME_ENABLED = volume_opts.fetch('enabled', false)
VOLUME_BUILD_TIMEOUT = (volume_opts['build_timeout'] || 120).to_i
TEST_VOLUME_SNAPSHOTS = volume_opts.fetch('test_snapshots', false)
CLEAN_UP_VOLUMES = volume_opts.fetch('cleanup', true)

#metering opts
metering_opts=configs['metering'] || {}
METERING_ENABLED = metering_opts.fetch('enabled', false)
METERING_SAMPLE_TIMEOUT = (metering_opts['sample_timeout'] || 360).to_i

#orchestration opts
orchestration_opts=configs['orchestration'] || {}
ORCHESTRATION_ENABLED = orchestration_opts.fetch('enabled', false)
STACK_CREATE_TIMEOUT = (orchestration_opts['stack_create_timeout'] || 120).to_i
CLEAN_UP_STACKS = orchestration_opts.fetch('cleanup', true)

FOG_VERSION=configs['fog_version']

TORPEDO_TEST_SUITE = Test::Unit::TestSuite.new("Torpedo")
module Torpedo

  class TorpedoTests
    def self.suite
      return TORPEDO_TEST_SUITE
    end
  end

  class Tasks < Thor

    desc "flavors", "Run flavors tests for the OSAPI."
    def flavors
      require 'torpedo/compute/flavors'
      TORPEDO_TEST_SUITE << Torpedo::Compute::Flavors.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "limits", "Run limits tests for the OSAPI."
    def limits
      require 'torpedo/compute/limits'
      TORPEDO_TEST_SUITE << Torpedo::Compute::Limits.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "images", "Run images tests for the OSAPI."
    def images
      require 'torpedo/compute/images'
      TORPEDO_TEST_SUITE << Torpedo::Compute::Images.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "servers", "Run servers tests for the OSAPI."
    def servers
      require 'torpedo/volume/volumes'
      require 'torpedo/compute/keypairs'
      require 'torpedo/compute/servers'
      require 'torpedo/metering/meters'
      require 'torpedo/cleanup'
      if VOLUME_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Volume::Volumes.suite
      end
      if KEYPAIR_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Compute::Keypairs.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Compute::Servers.suite
      if METERING_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Metering::Meters.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "keypairs", "Run keypair tests for the OSAPI."
    def keypairs
      require 'torpedo/compute/keypairs'
      require 'torpedo/cleanup'
      TORPEDO_TEST_SUITE << Torpedo::Compute::Keypairs.suite
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "volumes", "Run volume tests for the OSAPI."
    def volumes
      require 'torpedo/volume/volumes'
      require 'torpedo/cleanup'
      TORPEDO_TEST_SUITE << Torpedo::Volume::Volumes.suite
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "orchestration", "Run orchestration tests for the OSAPI."
    def orchestration
      require 'torpedo/compute/keypairs'
      require 'torpedo/orchestration/stacks'
      require 'torpedo/cleanup'
      if KEYPAIR_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Compute::Keypairs.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Orchestration::Stacks.suite
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "cleanup", "Clean up servers, images, volumes, etc."
    def cleanup
      require 'torpedo/cleanup'
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "all", "Run all tests."
    def all
      require 'torpedo/compute/keypairs'
      require 'torpedo/orchestration/stacks'
      require 'torpedo/compute/flavors'
      require 'torpedo/compute/limits'
      require 'torpedo/compute/images'
      require 'torpedo/volume/volumes'
      require 'torpedo/compute/servers'
      require 'torpedo/metering/meters'
      require 'torpedo/cleanup'
      if KEYPAIR_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Compute::Keypairs.suite
      end
      if ORCHESTRATION_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Orchestration::Stacks.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Compute::Flavors.suite
      TORPEDO_TEST_SUITE << Torpedo::Compute::Limits.suite
      TORPEDO_TEST_SUITE << Torpedo::Compute::Images.suite
      if VOLUME_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Volume::Volumes.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Compute::Servers.suite
      if METERING_ENABLED
        TORPEDO_TEST_SUITE << Torpedo::Metering::Meters.suite
      end
      TORPEDO_TEST_SUITE << Torpedo::Cleanup.suite
      exit Test::Unit::UI::Console::TestRunner.run(TorpedoTests).passed?
    end

    desc "ssh", "Test ssh connectivity on a server."
    method_options :ip_address => :string
    method_options :test_command => :string
    method_options :test_output => :string
    method_options :admin_password => :string
    def ssh(options=(options or {}))
      require 'torpedo/net_util'
      if Torpedo::NetUtil.ssh_test(options[:ip_address], nil, options[:test_command], options[:test_output], options[:admin_password]) then
        exit 0
      else
        exit 1
      end
    end

    desc "fire", "Fire away! (alias for all)"
    def fire
       invoke :all
    end

  end
end
